parsimonious/tests/test_grammar.py

```diff
@@ -493,3 +494,56 @@ def test_token_repr(self):
             self.assertEqual(u'<Token "💣">'.encode('utf-8'), t.__repr__())
         else:
             self.assertEqual(u'<Token "💣">', t.__repr__())
+
+
+def test_precedence_of_string_modifiers():
+    # r"strings", etc. should be parsed as a single literal, not r followed
+    # by a string literal.
+    g = Grammar(r"""
+        escaped_bell = r"\b"
+        r = "irrelevant"
+    """)
+    assert isinstance(g["escaped_bell"], Literal)
+    assert g["escaped_bell"].literal == "\\b"
+    with pytest.raises(ParseError):
+        g.parse("irrelevant\b")
+
+    g2 = Grammar(r"""
+        escaped_bell = r"\b"
+    """)
+    assert g2.parse("\\b")
+
+
+def test_binary_grammar():
+    g = Grammar(r"""
+        file = header body terminator
+        header = b"\xFF" length b"~"
+        length = ~rb"\d+"
+        body = ~b"[^\xFF]*"
+        terminator = b"\xFF"
+    """)
+    length = 22
+    assert g.parse(b"\xff22~" + (b"a" * 22) + b"\xff") is not None
+
+
+def test_inconsistent_string_types_in_grammar():
+    with pytest.raises(VisitationError) as e:
+        Grammar(r"""
+            foo = b"foo"
+            bar = "bar"
+        """)
+    assert str(e.value).startswith("BadGrammar:")
```

Comment:
You could probably check `e.original_class is BadGrammar` if you wanted. It feels an iota more foolproof to me.